core/src/errors.rs

```diff
@@ -1,5 +1,5 @@
 use core::fmt;
-use winter_utils::string::String;
+use winter_utils::string::*;
```

Comment:
I think we should import this and similar import form `crate::utils`, which, in turn, would be just re-exports from `miden-crypto`.

Reply:
I strongly disagree with that. IMO we have to stop re-exporting symbols on every crate and artificially adding these dependencies. The current code is using the form:

```
winter_utils <- miden_vm
          ^---- crypto
```

Meaning that updates to `miden_vm` and `crypto` can be done in parallel. Your suggestion changes the above to:

```
winter_utils <- crypto <- miden_vm
```

That serializes work, and forces updates to the repos to be done sequentially. This is in part why it is taking so much time to fix the clippy warnings, the complete chain is actually:

```
winter_utils <- crypto <- miden_vm <- miden_base <- miden_node
```

And the work is being done serially, and intermediary steps are confusing [e.g.](https://github.com/0xPolygonMiden/crypto/pull/280#issuecomment-1957410280).

IMO we should go to the opposite direction, and do what @bitwalker [proposed above](https://github.com/0xPolygonMiden/miden-vm/pull/1255#discussion_r1497808219), and do direct imports. That streamlines work, and has some other side benefits, namely:

- We don't have to chase re-exports on GH
    - Also, on the IDE. At least for me jump to definition doesn't always work (usually do to compilation errors), and I just can't know what the symbol means.
- This simplifies cargo's rendered docs, since the re-exports duplicate the docs

Reply:
> That serializes work, and forces updates to the repos to be done sequentially.

The serialization is there in either case because there are other serial dependencies. That is, if there is a change in `winter-crypto`, we cannot update `miden-vm` before first updating `miden-crypto` (even if each imports directly from `winter-crypto`) - otherwise we'd get conflicting imports and compilation will fail.

> We don't have to chase re-exports on GH
> * Also, on the IDE. At least for me jump to definition doesn't always work (usually do to compilation errors), and I just can't know what the symbol means.

I can't say that I've had similar issues with the IDE I'm using (VS Code) - but it is possible I didn't pay enough attention.

> This simplifies cargo's rendered docs, since the re-exports duplicate the docs.

I'm not sure this is a problem.

Overall, the main point of re-exporting core components is to simplify dependency management downstream. For example, in many cases, we can just import a single crate instead of importing several (or many) crates.

To take the example of `miden-crypto` - if we take the approach you've proposed to the extreme and don't re-export anything, any time we use `miden-crypto` we may also need to import `winter-crypto`, `winter-math`, and `winter-utils`. And then the user of `miden-crypto` needs to be aware of all these crates, which in my opinion, is an implementation detail of `miden-crypto`.

Basically, I don't see a reason why a user of `miden-crypto` needs to know that it is based `winter-crypto`, `winter-math` etc. Hiding these details also lets us update `miden-crypto` in the future in a less "breaking" way. For example, we may chose to get rid of `winter-utils` dependency and re-implement relevant pieces in `miden-crypto` (for performance or other reasons). If the user is only aware of `miden-crypto`, this change would be transparent to them. But if they rely on `winter-crypto` etc., this may be a breaking change.

Reply:
> The serialization is there in either case because there are other serial dependencies. That is, if there is a change in winter-crypto, we cannot update miden-vm before first updating miden-crypto (even if each imports directly from winter-crypto) - otherwise we'd get conflicting imports and compilation will fail.

- The conflicting imports should happen on major releases. Minor/patch releases should not have API breaking changes and the resolver should unify the dependencies, such that we shouldn't have conflicting imports. Major releases should be infrequent, and this should not be something that pops up frequently.
- There are workaround for the above: https://github.com/dtolnay/semver-trick , I don't have experience with this approach, but the point is that these breaking changes are not a given.
- This is missing the point I'm making. Namely that a change to `std`/`alloc` shouldn't require a release of winterfell to have a fix in miden-vm. That is what I'm calling an **artificial** dependency. If we just used `alloc` directly this would not have been an issue. Ofcourse, if we have to use winterfell, then that is another story, but this isn't what happened here with `Vec` and `String`.

> Overall, the main point of re-exporting core components is to simplify dependency management downstream. For example, in many cases, we can just import a single crate instead of importing several (or many) crates.

I'm not against re-exports for easy of use. But I do think this is a distinct conversation, namely about how to do a prelude. But I do think we shouldn't overlap our re-exports with the stdlib/core types.

---

I have update the code to use `crate::utils`

Reply:
I don't think re-exporting dependencies is always a bad thing, but I feel like re-exporting _anything_ from libcore/liballoc/libstd is non-idiomatic, and should be strongly discouraged. Part of the issue here is how these crates are being compiled to be no-std compatible. Crates that want to be no-std compatible should always compile as `#![no_std]`, add a feature called `std`, and then explicitly import things from libstd behind a `#[cfg(feature = "std")]` attribute.

Regarding the interactions with IDEs: You can configure rust-analyzer to automatically add imports for things from the standard library, and in the case of no-std crates, you can tell it to always prefer libcore/liballoc when doing so. This falls apart when you re-export dependencies, because the re-exports shadow the items in the standard library, so either it won't be able to auto-import for you, or it will import from the standard library and now you've got some imports using the standard library explicitly, and some which are using the re-exported aliases.

Re-exporting dependencies is supposed to be done very conservatively from what I've seen in the ecosystem, and is largely to avoid having downstream users of your crate need to go out of their way to manage the interaction of your crate with dependencies exposed in your public API that you don't control. However, when all of the crates in question are part of the same workspace, that doesn't apply - you control all of them, and if you depend on one, depending on others explicitly is, if anything, expected.

Take for example the `regex` crate, an oft-used example of a well architected Rust library. It builds on multiple other crates in a single workspace, most notably `regex_automata` and `regex_syntax`. Even though each depends on the next, they never re-export functionality from the crates underneath, instead the docs refer people to lower level crates explicitly if they want some functionality not provided by the higher-level crate. The `regex` crate depends on both `regex_automata` and `regex_syntax`, and `regex_automata` also depends on `regex_syntax` (and even has one of the `regex_syntax` types in its public API) - yet `regex_automata` does not re-export types from `regex_syntax`, and `regex` has both as explicit dependencies. 

I haven't dug around to see if BurntSushi provides a rationale, but they have carried those crates through a lot of evolution, both of Rust itself, and of the crates themselves, and if there had been a compelling reason to more tightly bind them together by re-exporting types up the dependency chain, they would have certainly done so by now.

Anyway, I'm mostly opinionated about the standard library stuff, and less so about regular dependencies, but figured I would at least provide another angle.

Reply:
> I don't think re-exporting dependencies is always a bad thing, but I feel like re-exporting _anything_ from libcore/liballoc/libstd is non-idiomatic, and should be strongly discouraged. Part of the issue here is how these crates are being compiled to be no-std compatible. Crates that want to be no-std compatible should always compile as `#![no_std]`, add a feature called `std`, and then explicitly import things from libstd behind a `#[cfg(feature = "std")]` attribute.

I think there are a couple of things here:

First, the way we handle no-std compilation is slightly different. We do have the `std` feature (which is usually enabled by default), but we don't use `#![no_std]` attribute. Instead, we avoid importing from `std` (try to use `core` whenever possible) and when we must use something from `std`, we use feature flags to import provide an alternative (e.g., from liballoc). I'm not sure why exactly we landed on this approach, but I don't see any issues with adding `#![no_std]` attribute to all crates to make it work as you described.

Second, the main reason for re-exporting from `liballoc` etc. is simpler code. For example, if we need to use `Vec` and `BTreeMap` in some module, we can write something like:
```Rust
use miden_core::utils::collections:*;
```
Instead of:
```Rust
[cfg(feature = "std")]
use std::{collections::BTreeMap, vec::Vec};

[cfg(not(feature = "std"))]
use alloc::{collections::BTreeMap, vec::Vec};
```
But if such re-exports are undesirable, we can:

1. Define `collections` (and other similar modules) in each crate and then the import would look like: `use crate::collections::*`. The actual [module](https://github.com/facebook/winterfell/blob/main/utils/core/src/collections.rs) is pretty small - so, this would lead to just a little bit of redundant code per crate.
2.  We could bite the bullet and write feature-gated imports explicitly in every modules. Code duplication also probably won't be huge (maybe a few hundred lines of extra code).

> Re-exporting dependencies is supposed to be done very conservatively from what I've seen in the ecosystem, and is largely to avoid having downstream users of your crate need to go out of their way to manage the interaction of your crate with dependencies exposed in your public API that you don't control. However, when all of the crates in question are part of the same workspace, that doesn't apply - you control all of them, and if you depend on one, depending on others explicitly is, if anything, expected.

I largely agree that extra imports are fine when we are talking about additional functionality. I think the tricky part is figuring out what is "core" and what is "additional". For example, if someone wants to import `miden-vm`, they shouldn't be forced to also import about a dozen other crates spanning miden and winterfell repos. This means that a lot of stuff the is defined in these crates needs to be re-exported via `miden-vm`. Similar logic applies to other crates (e.g., if I want to use `miden-assembly`, do I really also need to import `winter-math`, `winter-crypto`, and `miden-crypto` crates? And maybe the answer here is "yes" for `miden-crypto` but "no" for the other two).

We do have #570 still open where we discussed matters related to this - maybe it makes sense to take the discussion there (or better yet, probably close that issue and open a more relevant one).